Userland/Test.cpp

```diff
@@ -0,0 +1,540 @@
+        return make<Exists>(name);
+    } else if (arg == "-b") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-b expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::BlockDevice);
+    } else if (arg == "-c") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-c expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::CharacterDevice);
+    } else if (arg == "-d") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-d expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::Directory);
+    } else if (arg == "-f") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-f expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::Regular);
+    } else if (arg == "-h" || arg == "-L") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-h/-L expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::SymbolicLink);
+    } else if (arg == "-p") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-p expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::FIFO);
+    } else if (arg == "-S") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-S expects a file name");
+        return make<FileIsOfKind>(name, FileIsOfKind::Socket);
+    } else if (arg == "-r") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-r expects a file name");
+        return make<HasPermission>(name, HasPermission::Read);
+    } else if (arg == "-w") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-w expects a file name");
+        return make<HasPermission>(name, HasPermission::Write);
+    } else if (arg == "-x") {
+        StringView name = argv[++optind];
+        if (name.is_null())
+            fatal_error("-x expects a file name");
+        return make<HasPermission>(name, HasPermission::Execute);
+    } else if (arg == "-a") {
+        return nullptr;
+    } else if (arg == "-o") {
+        return nullptr;
+    } else if (arg == "-n") {
+        StringView value = argv[++optind];
+        return make<StringCompare>("", value, StringCompare::NotEqual);
+    } else if (arg == "-z") {
+        StringView value = argv[++optind];
+        return make<StringCompare>("", value, StringCompare::Equal);
+    } else if (arg == "-g" || arg == "-G" || arg == "-k" || arg == "-N" || arg == "-O" || arg == "-s") {
+        fatal_error("Unsupported operator \033[1m%s", argv[optind]);
+    } else {
+        // Try to read a binary op.
+        auto lhs = arg;
+        arg = argv[++optind];
```

Comment:
But what if the RHS of the binary operation is not a single argument, but a complex expression?

Reply:
how would that work?
AFAIK these binary ops take single args around them, e.g. `file -ef file`, you can't put anything except a single argument in its place...

Reply:
Ah, these aren't binary ops like "and" and "or", they're binary ops on two files, got it. Add a comment perhaps then?

Reply:
Sure.